features/opennms-es-rest/main-module/src/main/java/org/opennms/plugins/elasticsearch/rest/RestClientFactory.java

```diff
@@ -28,37 +28,130 @@
 
 package org.opennms.plugins.elasticsearch.rest;
 
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.opennms.core.utils.TimeoutTracker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.searchbox.action.Action;
 import io.searchbox.client.JestClient;
 import io.searchbox.client.JestClientFactory;
+import io.searchbox.client.JestResult;
+import io.searchbox.client.JestResultHandler;
 import io.searchbox.client.config.HttpClientConfig;
 
+/**
+ * This factory wraps the {@link JestClientFactory} to provide instances of
+ * {@link JestClient}.
+ */
 public class RestClientFactory {
-	
-	private JestClientFactory factory=null;
-	
-	
-	
+
+	private JestClientFactory factory = null;
+
+	private AtomicInteger m_timeout = new AtomicInteger(0);
+	private AtomicInteger m_retries = new AtomicInteger(0);
+
+	private static class OnmsJestClient implements JestClient {
+
+		private static final Logger LOG = LoggerFactory.getLogger(OnmsJestClient.class);
+
+		private final JestClient m_delegate;
+
+		private final int m_retries;
+
+		private final int m_timeout;
+
+		public OnmsJestClient(JestClient delegate, int timeout, int retries) {
+			m_delegate = delegate;
+			m_timeout = timeout;
+			m_retries = retries;
+		}
+
+		/**
+		 * Perform the REST operation and retry in case of exceptions.
+		 */
+		@Override
+		public <T extends JestResult> T execute(Action<T> clientRequest) throws IOException {
+			// 'strict-timeout' will enforce that the timeout time elapses between subsequent
+			// attempts even if the operation returns more quickly than the timeout
+			Map<String,Object> params = new HashMap<>();
+			params.put("strict-timeout", Boolean.TRUE);
+
+			TimeoutTracker timeoutTracker = new TimeoutTracker(params, m_retries, m_timeout);
+
+			for (timeoutTracker.reset(); timeoutTracker.shouldRetry(); timeoutTracker.nextAttempt()) {
+				timeoutTracker.startAttempt();
+				try {
+					return m_delegate.execute(clientRequest);
```

Comment:
The impact of what you're describing is limited to _doNotPersist_ events: any event that has an OpenNMS database ID will be indexed with that ID. If the record is indexed multiple times, you will just end up with multiple identical 'versions' of the same record but still, only a single record will be returned by ES queries.

It sounds like you're saying that we should not retry if we get a 'Read Timed Out' error but certainly in many cases when ES doesn't respond, it's likely that the records were not indexed properly. This sounds like a valid retry scenario.